File: src/api/connection.ts
import { useQuery, useQueryClient } from "@tanstack/react-query";
import { useEffect } from "react";
import { call, listen, type UnlistenFn } from "./transport";

/// What the mobile crate's `connection_state` command answers with.
///
/// The command belongs to issue #514 (the mobile client commands) and
/// does not exist yet; this file is the frontend's side of that
/// contract, written first so the connection banner has something to
/// render against. `state` is the five-way enum from the design spec;
/// `desktop` is the paired desktop's name from the pairing QR, null
/// while unpaired; `last_poll` is when the desktop last reported a
/// successful GitHub poll over `/v1/events`, ISO 8601, null before the
/// first one; `protocol_version` is what the desktop's `/v1/hello`
/// answered, absent or null until the phone has heard one. A report
/// from before the field existed reads the same as "not known".
interface ConnectionReport {
  state: "unpaired" | "connecting" | "connected" | "unreachable" | "revoked";
  desktop: string | null;
  last_poll: string | null;
  protocol_version?: number | null;
  /// True unless connected to a desktop the companion can drive: what
  /// the list's stale marker reads, and when the companion refuses
  /// write and destructive commands (`remote_call` rejects with a
  /// message naming the desktop and the reason).
  stale?: boolean;
}

/// The connection as the UI sees it.
///
/// `local` is the desktop app talking to its own Rust process: there is
/// no connection to report, and the banner renders nothing. `unknown` is
/// the mobile build before `connection_state` has answered -- including
/// when the command is not there, which is the case until #514 lands.
/// The rest are the command's own states, with the desktop's name
/// attached because every banner line needs it. `connected` also
/// carries the desktop's protocol version, because that is the one
/// state in which the phone would otherwise go on to issue commands: a
/// desktop older than `REQUIRED_PROTOCOL_VERSION` (`src/lib/protocol.ts`)
/// is reachable and yet must not be driven, and the banner says so.
export type ConnectionState =
  | { kind: "local" }
  | { kind: "unknown" }
  | { kind: "unpaired" }
  | {
      kind: "connected";
      desktop: string;
      lastPoll: string | null;
      /// Null while unknown; never null once the desktop has answered.
      protocolVersion: number | null;
      /// See `stale` on `ConnectionReport`. False in the ordinary
      /// connected case; true for a desktop that answers but must not
      /// be driven, such as one below the required protocol.
      stale: boolean;
    }
  | {
      kind: "connecting" | "unreachable" | "revoked";
      desktop: string;
      lastPoll: string | null;
      stale: boolean;
    };

/// Whether what the app is showing may be out of date.
///
/// The companion serves `get_cached` from its stored snapshot while the
/// desktop is away, so the list renders normally with no hint that its
/// rows are hours old -- which is exactly what this answers. `local` is
/// never stale: the desktop holds the data itself.
export function isStale(state: ConnectionState): boolean {
  switch (state.kind) {
    // The desktop holds the data itself, and an unpaired phone is
    // showing nothing to be stale about.
    case "local":
    case "unknown":
    case "unpaired":
      return false;
    default:
      return state.stale;
  }
}

/// How often the banner re-asks WITHOUT an event to go on.
///
/// The companion pushes `connection-state` on every change, so this is a
/// backstop rather than the mechanism: it catches a change whose event
/// was missed, which is most plausible around a suspension. Thirty
/// seconds costs the phone two webview wake-ups a minute instead of
/// twelve, and no state transition waits on it in practice.
const CONNECTION_POLL_MS = 30_000;

/// The event the companion emits on every connection change. Must match
/// `STATE_EVENT` in `src-mobile/src/connection.rs`.
const STATE_EVENT = "connection-state";

/// The mobile crate's answer. Through the transport, like every other
/// command: on the mobile build `remote.ts` invokes the companion's own
/// `connection_state` directly.
function connectionState(): Promise<ConnectionReport> {
  return call<ConnectionReport>("connection_state");
}

function fromReport(report: ConnectionReport): ConnectionState {
  if (report.state === "unpaired") return { kind: "unpaired" };
  // Absent means stale for every state but `connected`: a report from
  // before the field existed still describes a desktop the phone cannot
  // reach, and defaulting THAT to fresh would mark hours-old rows live.
  const stale = report.stale ?? report.state !== "connected";
  // A paired desktop always has a name -- it came from the QR -- but
  // the wire type allows null, and a banner reading "null is
  // unreachable" is worse than a generic noun.
  const desktop = report.desktop ?? "Desktop";
  if (report.state === "connected") {
    return {
      kind: "connected",
      desktop,
      lastPoll: report.last_poll,
      // Missing and null both mean "not known": the banner treats an
      // unknown version as fine, so a report that predates the field
      // does not turn into an update demand.
      protocolVersion: report.protocol_version ?? null,
      stale,
    };
  }
  return { kind: report.state, desktop, lastPoll: report.last_poll, stale };
}

const LOCAL: ConnectionState = { kind: "local" };

/// The desktop's answer: a constant. No query, and therefore no
/// QueryClient needed -- the banner mounts in every window and must
/// not make the desktop pay for a question only the phone asks.
function useLocalConnectionState(): ConnectionState {
  return LOCAL;
}

/// The phone's answer: the companion pushes every change, and a slow
/// poll underneath catches anything a missed event would have stranded.
///
/// It used to poll alone, every five seconds. `connection.rs` emits
/// `connection-state` on EVERY change (`STATE_EVENT`) and `remote.ts`
/// documents it -- "plus `connection-state` on every change, so `listen`
/// is Tauri's" -- and nothing listened. So the phone woke its webview
/// twelve times a minute for a value it was being handed for free, and
/// still showed a state up to five seconds stale.
///
/// The poll is kept, at a much longer interval, as a safety net: an
/// event dropped while the app was suspended would otherwise leave the
/// banner wrong until the next change, and a wrong connection banner is
/// what the stale marker depends on.
function useRemoteConnectionState(): ConnectionState {
  const client = useQueryClient();
  const { data } = useQuery({
    queryKey: ["connection-state"],
    queryFn: connectionState,
    refetchInterval: CONNECTION_POLL_MS,
    // A missing command rejects every time; retrying three times with
    // backoff before the next interval only delays the same answer.
    retry: false,
  });

  useEffect(() => {
    let unlisten: UnlistenFn | undefined;
    let live = true;
    try {
      // `listen` throws SYNCHRONOUSLY outside a Tauri runtime -- see the
      // note on the pass-throughs in `transport.ts`, which is why this
      // is a try/catch and not only a `.catch`. Without a runtime there
      // are no events to receive, and the poll below is the whole
      // mechanism; the banner is exactly as correct as it was before.
      void listen<ConnectionReport>(STATE_EVENT, (e) => {
        // Written straight into the cache rather than held in component
        // state: the query is the single source for this value, and two
        // copies would disagree the moment a poll landed between events.
        client.setQueryData(["connection-state"], e.payload);
      }).then(
        (off) => {
          // Unmounted before the subscription resolved: stop it now
          // rather than leaking a listener outliving the component.
          if (live) unlisten = off;
          else off();
        },
        () => {},
      );
    } catch {
      // As above: fall back to polling alone.
    }
    return () => {
      live = false;
      try {
        unlisten?.();
      } catch {
        // Tearing down a listener the runtime no longer knows about is
        // not worth failing an unmount over.
      }
    };
  }, [client]);

  return data === undefined ? { kind: "unknown" } : fromReport(data);
}

/// The current connection to the paired desktop.
///
/// Chosen once, at module load, from the build target: the target is a
/// compile-time constant, so this is one hook or the other for the life
/// of the process rather than a hook called conditionally.
export const useConnectionState: () => ConnectionState =
  import.meta.env.VITE_TARGET === "mobile" ? useRemoteConnectionState : useLocalConnectionState;

File: src-mobile/src/connection.rs
//! What the phone knows about its link to the desktop, as one value the
//! frontend polls (`connection_state`) and is told about (the
//! `connection-state` event).
//!
//! The states are the spec's five. `protocol_version` is what
//! `/v1/hello` reported on the last successful connect and is `None`
//! whenever the state is not `connected`, so a "desktop too old" banner
//! never reads a stale number. `last_poll` is when the desktop last
//! delivered a `prs-updated` (the snapshot on connect, or a poll result),
//! persisted with the snapshot so it survives a restart.

use chrono::{DateTime, Utc};
use serde::Serialize;
use std::sync::{Arc, Mutex};

use crate::client::PROTOCOL_VERSION;

/// Where events go. `lib.rs` implements it for the Tauri `AppHandle`
/// (`app.emit`); tests collect. `json` is the payload already
/// serialised, emitted verbatim.
pub trait EventSink: Send + Sync {
    fn emit(&self, name: &str, json: &str);
}

/// The Tauri event carrying a [`Report`] on every change, so the
/// frontend's transport can react without polling.
pub const STATE_EVENT: &str = "connection-state";

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum State {
    Unpaired,
    Connecting,
    Connected,
    Unreachable,
    Revoked,
}

/// The `connection_state` answer. Field names are the wire.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct Report {
    pub state: State,
    /// The paired desktop's name from the QR; `None` while unpaired.
    pub desktop: Option<String>,
    /// ISO 8601, or `None` before the first `prs-updated`.
    pub last_poll: Option<String>,
    /// From `/v1/hello`; `None` unless `state` is `connected`.
    pub protocol_version: Option<u32>,
    /// Whether what the phone shows may be out of date and actions are
    /// refused: anything but `connected` to a desktop speaking at least
    /// this app's protocol. The UI's stale marker reads this.
    pub stale: bool,
}

#[derive(Debug)]
struct Inner {
    state: State,
    desktop: Option<String>,
    last_poll: Option<DateTime<Utc>>,
    protocol_version: Option<u32>,
}

pub struct Connection {
    inner: Mutex<Inner>,
    sink: Arc<dyn EventSink>,
}

impl Connection {
    pub fn new(sink: Arc<dyn EventSink>) -> Self {
        Self {
            inner: Mutex::new(Inner {
                state: State::Unpaired,
                desktop: None,
                last_poll: None,
                protocol_version: None,
            }),
            sink,
        }
    }

    pub fn report(&self) -> Report {
        report_of(&self.inner.lock().unwrap_or_else(|e| e.into_inner()))
    }

    pub fn state(&self) -> State {
        self.inner.lock().unwrap_or_else(|e| e.into_inner()).state
    }

    /// Move to `state`. Any state but `Connected` clears
    /// `protocol_version`; use [`Connection::connected`] to enter
    /// `Connected` with one. Emits only on an actual change.
    pub fn set_state(&self, state: State) {
        self.update(|i| {
            if i.state == state {
                return false;
            }
            i.state = state;
            if state != State::Connected {
                i.protocol_version = None;
            }
            true
        });
    }

    /// `Connected`, with what the desktop said in `/v1/hello`.
    pub fn connected(&self, protocol_version: u32) {
        self.update(|i| {
            let changed =
                i.state != State::Connected || i.protocol_version != Some(protocol_version);
            i.state = State::Connected;
            i.protocol_version = Some(protocol_version);
            changed
        });
    }

    /// The desktop this phone is paired with (or not), set at load, at
    /// pairing, and at unpairing. Does not change `state`.
    pub fn set_desktop(&self, name: Option<String>, last_poll: Option<DateTime<Utc>>) {
        self.update(|i| {
            i.desktop = name;
            i.last_poll = last_poll;
            true
        });
    }

    pub fn mark_poll(&self, at: DateTime<Utc>) {
        self.update(|i| {
            i.last_poll = Some(at);
            true
        });
    }

    fn update(&self, f: impl FnOnce(&mut Inner) -> bool) {
        let report = {
            let mut inner = self.inner.lock().unwrap_or_else(|e| e.into_inner());
            if !f(&mut inner) {
                return;
            }
            report_of(&inner)
        };
        // Serialising a struct of scalars cannot fail.
        let json = serde_json::to_string(&report).expect("report serialises");
        self.sink.emit(STATE_EVENT, &json);
    }
}

/// Why write and destructive commands are refused right now, as the
/// clause that follows the desktop's name in the error, or `None` when
/// they are allowed.
fn blocked(i: &Inner) -> Option<&'static str> {
    match i.state {
        State::Unpaired => Some("is not paired with this phone"),
        State::Connecting => Some("is still being reached; try again in a moment"),
        State::Unreachable => Some("is unreachable; actions are disabled until it is back"),
        State::Revoked => Some("no longer recognises this phone; pair again"),
        State::Connected => match i.protocol_version {
            Some(v) if v >= PROTOCOL_VERSION => None,
            // #524's banner names the version and links the release.
            _ => Some("runs an older Headstate; update it before driving it from here"),
        },
    }
}

impl Connection {
    /// See [`blocked`].
    pub fn actions_blocked(&self) -> Option<&'static str> {
        blocked(&self.inner.lock().unwrap_or_else(|e| e.into_inner()))
    }
}

fn report_of(i: &Inner) -> Report {
    Report {
        state: i.state,
        desktop: i.desktop.clone(),
        last_poll: i
            .last_poll
            .map(|t| t.to_rfc3339_opts(chrono::SecondsFormat::Secs, true)),
        protocol_version: i.protocol_version,
        stale: blocked(i).is_some(),
    }
}

#[cfg(test)]
pub(crate) mod tests {
    use super::*;
    use std::sync::Mutex;

    /// Collects every emitted event, for tests across the crate.
    #[derive(Default)]
    pub(crate) struct Recorder {
        pub events: Mutex<Vec<(String, String)>>,
    }

    impl Recorder {
        pub(crate) fn names(&self) -> Vec<String> {
            self.events
                .lock()
                .unwrap()
                .iter()
                .map(|(n, _)| n.clone())
                .collect()
        }
        pub(crate) fn last(&self, name: &str) -> Option<String> {
            self.events
                .lock()
                .unwrap()
                .iter()
                .rev()
                .find(|(n, _)| n == name)
                .map(|(_, j)| j.clone())
        }
    }

    impl EventSink for Recorder {
        fn emit(&self, name: &str, json: &str) {
            self.events
                .lock()
                .unwrap()
                .push((name.to_string(), json.to_string()));
        }
    }

    fn conn() -> (Arc<Recorder>, Connection) {
        let rec = Arc::new(Recorder::default());
        (rec.clone(), Connection::new(rec))
    }

    #[test]
    fn the_report_is_the_wire_shape_the_frontend_expects() {
        let (_, c) = conn();
        let json = serde_json::to_value(c.report()).unwrap();
        assert_eq!(
            json,
            serde_json::json!({
                "state": "unpaired",
                "desktop": null,
                "last_poll": null,
                "protocol_version": null,
                "stale": true
            })
        );
        c.set_desktop(Some("octocat's laptop".into()), None);
        c.connected(1);
        c.mark_poll("2026-09-05T12:00:00Z".parse().unwrap());
        assert_eq!(
            serde_json::to_value(c.report()).unwrap(),
            serde_json::json!({
                "state": "connected",
                "desktop": "octocat's laptop",
                "last_poll": "2026-09-05T12:00:00Z",
                "protocol_version": 1,
                "stale": false
            })
        );
    }

    #[test]
    fn stale_and_blocked_track_state_and_protocol() {
        let (_, c) = conn();
        assert!(c.report().stale);
        assert!(c.actions_blocked().unwrap().contains("not paired"));
        c.connected(PROTOCOL_VERSION);
        assert!(!c.report().stale);
        assert_eq!(c.actions_blocked(), None);
        c.connected(PROTOCOL_VERSION + 1);
        assert_eq!(c.actions_blocked(), None, "a newer desktop is fine");
        c.connected(PROTOCOL_VERSION - 1);
        assert!(c.report().stale);
        assert!(c.actions_blocked().unwrap().contains("older Headstate"));
        c.set_state(State::Unreachable);
        assert!(c.actions_blocked().unwrap().contains("unreachable"));
        c.set_state(State::Revoked);
        assert!(c.actions_blocked().unwrap().contains("pair again"));
    }

    #[test]
    fn protocol_version_is_null_in_every_state_but_connected() {
        let (_, c) = conn();
        c.connected(1);
        assert_eq!(c.report().protocol_version, Some(1));
        for s in [
            State::Connecting,
            State::Unreachable,
            State::Revoked,
            State::Unpaired,
        ] {
            c.connected(1);
            c.set_state(s);
            assert_eq!(c.report().protocol_version, None, "{s:?}");
            assert_eq!(c.state(), s);
        }
    }

    #[test]
    fn every_change_is_emitted_and_a_no_op_is_not() {
        let (rec, c) = conn();
        c.set_state(State::Connecting);
        c.set_state(State::Connecting);
        c.connected(1);
        c.connected(1);
        assert_eq!(rec.names(), vec![STATE_EVENT, STATE_EVENT]);
        let last: serde_json::Value =
            serde_json::from_str(&rec.last(STATE_EVENT).unwrap()).unwrap();
        assert_eq!(last["state"], "connected");
    }
}
